src/chunkserver/scan_manager.cpp

```diff
@@ -239,47 +208,85 @@ void ScanManager::GenScanJobs(ScanKey key) {
+    while (iter != job->chunkMap.end()) {
+        // check chunk version
+        auto csChunkFile = iter->second;
+        if (csChunkFile->GetChunkFileMetaPage().version !=
+            FORMAT_VERSION_V2) {
+            iter++;
+        } else {
+            // split scan chunk request
+            job->currentChunkId = iter->first;
+            uint32_t currentOffset = 0;
+            bool scanChunkMetaPage = true;
+            while (currentOffset < chunkSize_) {
+                // check is leader, if not cancel the job
+                if (!nodePtr->IsLeaderTerm()) {
+                    CancelScanJob(job->poolId, job->id);
+                    return -1;
+                }
 
-        // construct scan task
-        ChunkResponse *response = new ChunkResponse();
-        ChunkRequest *request = new ChunkRequest();
-        request->set_optype(CHUNK_OP_TYPE::CHUNK_OP_SCAN);
-        request->set_logicpoolid(job->poolId);
-        request->set_copysetid(job->id);
-        request->set_chunkid(job->currentChunkId);
-        request->set_offset(currentOffset);
-        request->set_size(scanSize_);
-        request->set_sendscanmaptimeoutms(timeoutMs_);
-        request->set_sendscanmapretrytimes(retry_);
-        request->set_sendscanmapretryintervalus(retryIntervalUs_);
-        ScanChunkClosure *done = new ScanChunkClosure(request, response);
-        std::shared_ptr<ScanChunkRequest> req =
-            std::make_shared<ScanChunkRequest>(nodePtr, this, request,
-                                               response, done);
-        req->Process();
-        currentOffset += scanSize_;
-        // wait for scan task finished
-        uint32_t retry = retry_;
-        while (!job->isFinished && retry > 0) {
-            scanTaskWaitInterval_.WaitForNextExcution();
-            retry--;
+                // Init job
+                job->taskLock.WRLock();
+                job->task.localMap.Clear();
+                job->task.followerMap.clear();
+                job->task.waitingNum = 3;
+                job->task.chunkId = job->currentChunkId;
+                job->task.offset = currentOffset;
+                job->taskLock.Unlock();
+                job->currentOffset = currentOffset;
+                job->isFinished = false;
+
+                // construct scan task
+                ChunkResponse *response = new ChunkResponse();
+                ChunkRequest *request = new ChunkRequest();
+                request->set_optype(CHUNK_OP_TYPE::CHUNK_OP_SCAN);
+                request->set_logicpoolid(job->poolId);
+                request->set_copysetid(job->id);
+                request->set_chunkid(job->currentChunkId);
+                request->set_offset(currentOffset);
+                request->set_sendscanmaptimeoutms(timeoutMs_);
+                request->set_sendscanmapretrytimes(retry_);
+                request->set_sendscanmapretryintervalus(retryIntervalUs_);
+                if (scanChunkMetaPage) {
```

Comment:
for scan meta page, you can just set `request->set_readmetapage(true)`, ellipsis `set_offset`/`set_size` and member variable `chunkMetaPageSize_`

Reply:
> for scan meta page, you can just set `request->set_readmetapage(true)`, ellipsis `set_offset`/`set_size` and member variable `chunkMetaPageSize_`

'chunkMetaPageSize_' is not needed now, but  offset and length is better to set is error occur can log more info.